clusterous/cluster.py

```diff
@@ -643,6 +640,161 @@ def get_cluster_spec(self):
+
+        ssh = paramiko.SSHClient()
+        ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        ssh.connect(hostname = nat_public_ip,
+                    username = 'ec2-user',
+                    key_filename = os.path.expanduser(self._config['key_file']))
+        cmd = 'sudo iptables -t nat -A PREROUTING -p tcp --dport 22000 -j DNAT --to-destination {0}:22'.format(controller_private_ip)
+        # Using 'invoke_shell' because it is 'sudo' cmd
+        channel = ssh.invoke_shell()
+        channel.send('{0}\n'.format(cmd))
+        time.sleep(3)
+        output = channel.recv(99999)
+        ssh.close()
+        
+        if '000 -j DNAT' not in output:
+            message = 'Error: Setting SSH port forwarding on NAT instance.'
+            return (False, message)
+
+        return (True, '')
+
+    def _tag_resource(self, resource_name):
+        return {'Name': self.cluster_name + '-'+str(resource_name), defaults.instance_tag_key: self.cluster_name}
+    
+    def _create_vpc(self, vpc_id = None):
+        vpc = None
+
+        if vpc_id is None:
+            vpc_name_full = "{0}-vpc".format(self.cluster_name)
+            
+            # Get list of VPCs
+            vpcs = self.vpc_conn.get_all_vpcs()
+            for i in vpcs:
+                if i.tags.get('Name','') == vpc_name_full:
+                    vpc = i
+        
+            if vpc is None:
+                # Create VPC
+                vpc = self.vpc_conn.create_vpc('10.2.0.0/16')
+                vpc.add_tags(self._tag_resource('vpc'))
+                self.vpc_conn.modify_vpc_attribute(vpc.id, enable_dns_support=True)
+                self.vpc_conn.modify_vpc_attribute(vpc.id, enable_dns_hostnames=True)
+        
+                # Create network acl
+                network_acl = self.vpc_conn.create_network_acl(vpc.id)
+                network_acl.add_tags(self._tag_resource('acl'))
+
+        else:
+            vpc = self.vpc_conn.get_all_vpcs(vpc_ids=[vpc_id])[0]
+    
+        return vpc
+    
+    def _create_private_sg(self, sg_name):
+        security_group = None
+        security_group_name_full = "{0}-{1}".format(self.cluster_name, sg_name)
+        security_groups = self.vpc_conn.get_all_security_groups()
+        for i in security_groups:
+            if i.tags.get('Name','') == security_group_name_full:
+                security_group = i
+    
+        if security_group is None:
+            # Private Security group
+            security_group = self.vpc_conn.create_security_group(self.cluster_name+"-{0}".format(sg_name), 'Private security group for ' + self.cluster_name, self.vpc.id)
+            security_group.add_tags(self._tag_resource(sg_name))
+            security_group.authorize(ip_protocol='-1', from_port=0, to_port=65535, cidr_ip='0.0.0.0/0')
+        
+        return security_group
+
+    def _create_public_sg(self, sg_name):
```

Comment:
The _create_private_sg and _create_public_sg methods are identical, except for 4 lines. Perhaps they could be combined? The public sg authorisation could probably go in a different method.